### PR TITLE
fixes #20681; add efSkipFieldVisibilityCheck to skip check

### LINKED ISSUE #20681: Can not use nim 2's new default instantiation with any object type with a `DateTime` field
### What happened?

It is currently impossible to instantiate a DateTime object at compile time. That makes it impossible to make use of nim 2's new feature of [default instantiation](https://github.com/nim-lang/Nim/pull/20480) for any object with fields of the `DateTime` type.

These are the ways I tried so far:

Attempt 1:
```nim
import std/times

type A = object
  d: DateTime = DateTime()

let x = default(A)
```
Attempt 2:
```nim
import std/times

type A = object
  d: DateTime = dateTime(2020, 1.Month, 1)

let x = default(A)
```
Attempt 3:
```nim
import std/times

type A = object
  d: DateTime = now()

let x = default(A)
```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-23
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: fa1606c6d6799823de51f7a273ae5da2f3fa6e3a
active boot switches: -d:release

### Current Standard Output Logs

```shell
Attempt 1:
Error: the field 'nanosecond' is not accessible.


Attempt 2:
/.choosenim/toolchains/nim-#devel/lib/pure/times.nim(1366, 6) Error: cannot evaluate at compile time: localInstance


Attempt 3:
/.choosenim/toolchains/nim-#devel/lib/pure/times.nim(971, 13) Error: cannot 'importc' variable at compile time; clock_gettime
```

### Expected Standard Output Logs

```shell
No Error message as above
```

### Possible Solution

Suggestions:

1. Provide a separate proc that can solely be used to instantiate some dummy value for DateTime at compileTime.
2. Make the fields of DateTime public so that one can use `DateTime()` to instantiate a DateTime themselves with dummy values

### Additional Information

I believe this to be pretty relevant in particular for the feature provided by https://github.com/nim-lang/Nim/pull/20480 , as types with DateTimes aren't all that rare. Unless such a feature can be provided I effectively am unable to make use of it, as I must use ElegantBeef's `constructor` package for all types with DateTime fields and when I have that dependency anyway I might as well use it in general.

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20681

Constant object construction has been checked and `inlined` already, there is no need to `sem` it again, especially for default fields.

fixes
```nim
import std/tables
block:
  type TokenData = object
    campaignMemberships = Table[string, string]()

  let x = default(TokenData)
  doAssert x.campaignMemberships.len == 0
```

### CODE REVIEW

File: compiler/semobjconstr.nim [hunk 1]
```diff
@@ -380,6 +380,8 @@ proc defaultConstructionError(c: PContext, t: PType, info: TLineInfo) =
     assert false, "Must not enter here."
 
 proc semObjConstr(c: PContext, n: PNode, flags: TExprFlags; expectedType: PType = nil): PNode =
+  if nfUseDefaultField in n.flags:
+    return n
```
Comment (Araq) [inline compiler/semobjconstr.nim]: Why do we care? It's `semObjConstr`'s job to semcheck an object construction. It's not its job to "cache" the value or to detect if it already was semchecked. Why was it already semchecked? Where does that happen?

Comment (ringabout) [reply compiler/semobjconstr.nim]:
```nim
import std/tables
block:
  type TokenData = object
    campaignMemberships = Table[string, string]()
 ```

`Table[string, string]()` is evaluated at the compile time. Before it is passed to `semObjConstr`, it has been inlined to `Table(data: [], state: 0)` by `getConstExpr`. But `data` is a private field from another module. If it is checked here, the `semObjConst` will report `the field '$1' is not accessible.`. I think we should allow `semObjectConst` to ignore constant object construction.

Comment (ringabout) [reply compiler/semobjconstr.nim]: Sure, I will add `efSkipFieldVisibilityCheck` flag instead.

Comment (ringabout) [reply compiler/semobjconstr.nim]: Tried to semcheck the constant fields, but there is something wrong with the construction or type:

```
Table[string, string]()
------semConstExpr-------
Table(data: [], counter: 0)
```

```
tobject_default_value.nim(437, 13) template/generic instantiation of `main` from here
tobject_default_value.nim(424, 50) Error: type mismatch: got <array[0..-1, empty]> but expected 'KeyValuePairSeq[system.string, system.string]'
```
`data` is a seq type, but VM returns an empty array. It cannot pass `semconstr`.

```json
{
      "kind": "nkSym",
      "typ": "tyGenericInst",
      "sym": {
        "kind": "skField",
        "name": "data",
        "id": 503316488,
        "position": 0,
        "flags": ["sfFromGeneric"]
      }
    }, {
      "kind": "nkBracket",
      "typ": "tyGenericInst"
    }]
  }
```

Comment (ringabout) [reply compiler/semobjconstr.nim]: Now I add efSkipFieldVisibilityCheck and fix the type of nkBracket.

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -3075,7 +3075,10 @@ proc semExpr(c: PContext, n: PNode, flags: TExprFlags = {}, expectedType: PType
     of paTupleFields: result = semTupleFieldsConstr(c, n, flags, expectedType)
     of paSingle: result = semExpr(c, n[0], flags, expectedType)
   of nkCurly: result = semSetConstr(c, n, expectedType)
-  of nkBracket: result = semArrayConstr(c, n, flags, expectedType)
+  of nkBracket:
+    result = semArrayConstr(c, n, flags, expectedType)
+    if n.typ != nil and n.typ.skipTypes(abstractRange).kind == tySequence:
+      result.typ = n.typ # keeps tySequence type for []
```
Comment (ringabout) [inline compiler/semexprs.nim]: It is not correct. I will add a `semSeqConstr` soon.

**Comment by ringabout** [comment]
It seems that there is probably nothing wrong with this patch.

I have spent a bad time debugging the macros of loopfusion, I think these changes may work.

image

With this patch, the type of `tySequence` is not erased for `nkBracket`, which causes loopfusion to stop working.

Am I correct or do I miss something?

**Comment by ringabout** [comment]
wait for https://github.com/nim-lang/Nim/pull/20662

**Pull request merged**